Step3

dongyyy opened the pull request:
feat(step3): 문자열 계산기 구현하였습니다.
너무 늦어서 죄송해요. 천천히 봐주시면 감사하겠습니다!

dongyyy committed:
feat: 버튼 0~9 누르면 0~9 가 화면에 보인다. (0을 여러번 누르면 테스트 실패한다.)

dongyyy committed:
feat(step2): 모듈 분리

dongyyy committed:
2단계 피드백 반영

- 계산기의 도메인 모듈인걸 알기 쉽도록 패키지명을 edu.nextstep.camp.calculator.domain으로 변경

- jvm모듈이 아닌 코틀린 순수 모듈로 변경 - id 'kotlin'

- kotlin_version 변수의 값을 이용하여 kotlin-gradle-plugin dependencies 버전 할당

- 실패하는 테스트 케이스 삭제

dongyyy committed:
feat(step3): 문자열 계산기

dongyyy committed:
Merge branch 'dongyyy' into step3

namjackson commented on app/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,99 @@
+        val expected = 5
+
+        //then
+        assertThat(Calculator.evaluate(expression, delimiter)).isEqualTo(expected)
```
given when then 표기법으로 작성해 주셨네요 👍 
```
//when
val result = Calculator.evaluate(expression, delimiter)
//then
assertThat(result).isEqualTo(expected)
```
when, then을 분리하시면 좀더 좋을거같아요!

namjackson commented on app/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,99 @@
+import org.junit.Assert.assertThrows
+import org.junit.Test
+
+class CalculatorTest {
```
Calculator 클래스가 도메인 모듈로 옮겨갔습니다! 
해당하는 Test 코드도 도메인 모듈로 옮겨주세요!

namjackson commented on app/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,99 @@
+        assertThat(Calculator.evaluate(expression, delimiter)).isEqualTo(expected)
+    }
+
+    //입력값이 null이거나 빈 공백 문자일 경우 IllegalArgumentException throw
```
주석을 사용하기보다는 테스트 함수명을 활용해주세요! 
fun `한글_작성이_가능합니다`()
한글 함수명이 아니더라도, 구체적인 테스트케이스명을 적어주는게 좋을거같아요!

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+            "+" -> plus(operandFirst, operandSecond)
+            "-" -> minus(operandFirst, operandSecond)
+            "*" -> multiply(operandFirst, operandSecond)
+            "/" -> divide(operandFirst, operandSecond)
```
연산자라는 단위의 문자열이 여러종류가 있네요!
enum class를 활용해서 묶어보면 어떨까요?

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+
+        val inputList: List<String> = input.split(delimiter)
+
+        inputList.forEach { println(it) }
```
개발중에 작성한 불필요한 코드는 삭제해주세요!

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+        return result
+    }
+
+    fun requiredInput(input: String?) {
```
기본적으로 외부에서 사용하지 않는 함수는 Private으로 선언하는걸 지향합니다! 
private으로 인해 테스트가 힘들땐, 해당 역할을 답당하는 클래스를 분리해보는것도 좋은 방법입니다!

추가적으로 파라미터로 들어오는 표현식에 대한 검증이 부족한거같아요! 
"1+1" , "a + 123", "1 + + 1" 등등 수 많은 오류케이스가 존재하는데 
검증 테스트 및 검증단계가 부족한거같아요! 

문자열을 검증하는 역할을 담당하는 클래스와 테스트를 별도로 생성해보면 어떨까요?

namjackson replied:
Input이라는 이름은 매우 광번위하고, 어떠한 값이 input되어 들어오는거지
Input을 의미하는 값이 들어오는것이 아닙니다.
rawExpression 같은 의미의 이름이 조금더 어울릴거 같아요!

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+            "-" -> minus(operandFirst, operandSecond)
+            "*" -> multiply(operandFirst, operandSecond)
+            "/" -> divide(operandFirst, operandSecond)
+            else -> throw IllegalArgumentException() //사칙연산 기호가 아닌 경우 IllegalArgumentException throw
```
주석보다는 Exception내의 메시지를 포함해주셔도 좋을거같아요! 
`IllegalArgumentException("사칙연산 기호가 아닙니다.")`

namjackson commented on app/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,99 @@
+        val expression = " "
+
+        //then
+        assertThrows(IllegalArgumentException::class.java){Calculator.requiredInput(expression)}
```
커밋전에 항상 린트를 활용하여 자동정렬을 잊지말아주세요! 
(Cmd + Alt + L)

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,47 @@
+
+object Calculator {
+
+    fun evaluate(input: String, delimiter: String): Int {
```
delimiter를 입력으로 받는 방식도 좋지만,
현재 요구사항에서는 명확하게 " "으로 명시되어있습니다! 
상수로 선언하여 사용하면 좋을거같아요!

dongyyy replied:
"MVC - 단위 테스트, UI 테스트" 의 4단계 "계산기 UI" 를 나중에 진행하고 
먼저 MVP 과제를 진행해도 될까요?

namjackson reviewed (changes requested):
3단계 미션 고생하셨습니다! 
몇가지 코멘트 남겼으니 확인해주세요!
변경사항 적용하시고, 다시 리뷰요청 해주세요!

dongyyy committed:
when, then을 분리

dongyyy committed:
도메인 모듈로 Calculator Test 코드 옮기기

Calculator 클래스가 도메인 모듈로 옮겨갔습니다!
해당하는 Test 코드도 도메인 모듈로 옮겨주세요!

dongyyy committed:
주석을 사용하기보다는 테스트 함수명을 활용하도록 수정

dongyyy committed:
개발중에 작성한 불필요한 코드 삭제

dongyyy committed:
연산자 문자열 enum class로 묶기

dongyyy committed:
CalculatorTest failed로  수정 원복

dongyyy committed:
수식 유효성 검증 클래스를 분리 & 테스트 생성, 네이밍 변경(Input-> rawExpression)

dongyyy committed:
주석보다는 Exception내의 메시지를 포함

dongyyy committed:
린트를 활용하여 자동정렬 & 클래스 정리

dongyyy committed:
delimiter 요구사항에 맞게 상수로 선언

delimiter를 입력으로 받는 방식도 좋지만,
현재 요구사항에서는 명확하게 " "으로 명시되어있습니다!
상수로 선언하여 사용하면 좋을거같아요!

namjackson commented on domain/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,177 @@
+        //given
+        val expression = "2 + 3"
+        val delimiter = " "
+        val expected = 5
```
given, when, then으로 표현을 잘해주셨습니다!

테스트 케이스가 많아지고, 복잡해질 경우, 단순 `덧셈 테스트`라는 테스트 케이스 명은
모호한 케이스명이 될 가능성이 큽니다. 

"2 + 3"의 연산식을 계산하면, 결과가 "5"가 나와야한다 와 같이 구체적인 테스트 명이 관리하기에도, 직관적으로도 좋지않을까요?

namjackson commented on domain/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,177 @@
+        val expected = 5
+
+        //when
+        val result = Calculator.evaluate(expression, delimiter)
```
리팩토링과정에 따라서, 테스트 코드가 업데이트가 안되었습니다ㅜㅜ! 
현업에서도 이런 케이스가 자주일어날수 있는데요,
리팩토링 과정 이후, 테스트로 검증을 하는 습관이 있으면 좋을거같아요!

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/RawExpressionCheck.kt:
```diff
@@ -0,0 +1,16 @@
+
+class RawExpressionCheck {
+
+    fun isNullOrBlankCheck(rawExpression: String?) {
```
isXXX 라는 함수 명은 보통 Boolean을 리턴해주는데요, 
혼돈이 있을거 같아요,
`checkNullOrBlank, checkNumber ` 와 같은 이름은 어떨까요?

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+    private fun divide(operandFirst: Int, operandSecond: Int) =
+        operandFirst / operandSecond
+
+    enum class Operators(val operator: String) {
```
Operator을 enum 클래스를 작성해주셨네요 👍 
- enum class는 복수를 지칭하는 클래스가 아니라서, Operator 클래스명이 더 어울릴거 같아요! 
- `val operator: String`라는 변수명은 Operator.XXX.operator 클래스명과 중복인거 같아요! (value, token 등과 같은 이름은 어떨까요? 
- enum 클래스를 좀더 활용해보면 어떨까요? 바로위의 plus,minus등의 연산 함수는 Operator클래스에서 각각의 연산자 내에서 수행되면 좋을거같아요! 
https://kotlinlang.org/docs/enum-classes.html#anonymous-classes

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -0,0 +1,59 @@
+    private const val delimiter = " "
+
+    fun evaluate(expression: String): Int {
+        rewExpressionCheck.isNullOrBlankCheck(expression)
```
expression에 대한 검증을 여러과정에서 여러번 수행이 되는데요, 
문자열검증과정은 정규표현식(숫자+(연산자+숫자))을 활용하는 방법을 많이 사용하곤합니다!

dongyyy replied:
앗.. 이건 힌트를 좀 더 주실 수 있을까요?

namjackson commented on domain/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,177 @@
+
+        //then
+        assertThrows(IllegalArgumentException::class.java) {
+            rawExpressionCheck.isNullOrBlankCheck(expression)
```
CalculatorTest 테스트에서는, Calculator의 테스트를 집중해주세요!
`Calculator.evaluate(expression)`

CalculatorTest에서 RawExpressionCheck클래스 테스트 코드가 있는데, 
테스트 코드 또한 클래스단위로 나눠주시는게, 테스트 코드 관리에 수월할거 같아요!
rawExpressionCheck 테스트의 경우, RawExpressionCheck클래스의 테스트 코드를 따로 작성해주세요!

namjackson commented on domain/src/test/java/camp/nextstep/camp/calculator/domain/CalculatorTest.kt:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @Test
+    fun `파라미터로 들어오는 표현식에 대한 테스트 3`() {
```
테스트 1, 테스트 2의 네이밍은 작성자의 의도를
다른 사람에게 전달하기 매우 힘들기 때문에, 지양해주시는 것이 좋습니다. 

보다 구체적인 케이스명을 작성하면, 구현코드를 확인하지않고도 어떤테스트인지 알수있을거같아요! 

```
- 피연산자에 숫자가 아닌 값이 아닌 표현식이 경우, 에러가 발생해야한다.  
- "a + 123" 표현식이 경우, 에러가 발생해야한다.  
```

namjackson reviewed (changes requested):
피드백 반영하시느라 고생하셨습니다. 
추가로 몇가지 고민해볼법한 내용을 남겼습니다!
확인해줏세요!

namjackson commented:
주차별 미션 수행 순서는 상관없습니다!
MVC->MVP로 변경되기때문에, 앞선 주차 과제를 먼저 해주시는게 좋긴한데,
주차별 과제들을 동시에 진행하셔도 됩니다! 

개인적으로는 단순히 단계별로 의무적으로 과제를 하시기보단, 
동영님이 많은 부분을 얻을수 있도록 의미있게, 진행 해주시는게 좋을거같아요! 

교육 기간내에만 해주시면 상관없을거 같아요! ( 교육 종료 후에도 언제든 리뷰를 해주시는걸로 알고있는데, 교육과정 이후에는 리뷰가 늦을수도 있다고하시니 참고하세요!)

dongyyy committed:
피드백 반영

모호한 케이스명 수정
리팩토링 중, 테스트 코드가 업데이트 안됨
혼동을 줄 수 있는 함수명 수정(isXXX 라는 함수 명은 보통 Boolean을 리턴)
enum 클래스를 좀 더 활용(enum class명 수정, 클래스명과 변수명 중복 수정, anonymous-classes 사용)
rawExpressionCheck 테스트의 경우, RawExpressionCheck클래스의 테스트 코드를 따로 작성
테스트 1 , 테스트 2 메서드의 네이밍을 보다 구체적인 케이스명으로 수정

namjackson commented on domain/src/test/java/camp/nextstep/camp/calculator/domain/RawExpressionCheckTest.kt:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Test
+    fun `입력값이 null일 경우 IllegalArgumentException throw 테스트`() {
```
구체적인 테스트 케이스명 👍 👍

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -30,30 +30,32 @@ object Calculator {
+            Plus.value -> Plus.operate(operandFirst, operandSecond)
+            Minus.value -> Minus.operate(operandFirst, operandSecond)
+            Multiply.value -> Multiply.operate(operandFirst, operandSecond)
+            Divide.value -> Divide.operate(operandFirst, operandSecond)
```
operate로 각각의 연산에 대한 내용을 적의해주셨네요! 👍 

이렇게 되면 사실, when문은 산관없을거같아요. 
operator. operate(operandFirst, operandSecond) 이라고 호출하면 되지않을까요?
계산기의 입장에선, 연산이 되는지가 중요하지, 어떤 연산을 하는지는 중요하지않은거같아요!

dongyyy replied:
when 문을 안쓰고 구현할 수 있다는 말씀인가요?
오래 생각해봤는데 모르겠네요.. 어떻게 구현해야할까요?

namjackson commented on domain/src/main/java/camp/nextstep/camp/calculator/domain/Calculator.kt:
```diff
@@ -9,15 +9,15 @@ object Calculator {
 
     fun evaluate(expression: String): Int {
-        rewExpressionCheck.isNullOrBlankCheck(expression)
+        rewExpressionCheck.checkNullOrBlank(expression)
```
질문 주신걸 지금 봤네요 ㅠㅠ 
정규표현식은 Regex를 사용하여,
regex.matches(expression)으로, 해당 문자열이 계산식의 규칙 맞는지 확인할수 있어요! 

`\\d+(연산자문자열\\d+)+` 이런식으로 구현할수 있긴합니다!
d는 숫자를 가르키고, +는 앞의 표현식이 여러번 올수도 있다는걸 의미해요. 

정규표현식이 사실 정답은 아니지만,
사용해보시지 않으셧으면 이번 기회에 사용해도 좋을거같아요!

namjackson reviewed (approved):
리뷰 적용 잘해주셨네요, 고생하셨습니다! 
몇가지 참고될만한 내용,질문에 대한 답변 코멘트로 남겼으니,  확인해주세요! 
간단한 변경사항은 다음 미션에서 적용해주세요!

namjackson commented:
혹시 제가 질문에 대해 답변을 놓치면,
편하게 부담없이 dm을 주셔도 좋아요!!
